### PR TITLE
refactor: reusable scripts for local setup

### PR DESCRIPTION
**Pull request by yrong**

### CODE REVIEW

**Comment by vgeddes** [comment]
Very nice! @claravanstaden will definitely appreciate this.

**Comment by claravanstaden** [comment]
This looks great! I'll check it out and play with it for a bit, then we can merge. 😄

**Review by vgeddes — COMMENTED** [review]
@yrong Do you have an M1 Macbook by any chance? I want to see if it is possible to get these  scripts running on MacOS, so we can test without having to run a separate Ubuntu VM / Instance.

If not, we can try handle the macbook issue in another PR.

File: core/packages/test/scripts/set-env.sh
Comment (claravanstaden) [inline core/packages/test/scripts/set-env.sh:64]: I'm reaching this condition, and thinking we should run `scripts/build-binary.sh` here instead of quitting.

**Review by claravanstaden — CHANGES_REQUESTED** [review]
Hey @yrong! Thanks for this, super useful! 😄 I tested this out on my dev machine. I ran start-services.sh as usual:

```
/scripts/start-services.sh 
/home/ubuntu/projects/ron-snowbridge/relaychain/target/release/polkadot does not exist,run scripts/build-binary.sh first
Terminated
```

I am wondering if you encounter this condition, can't you run the `build-binary.sh` as a fix and then continue the script? Instead of having the user having to run that first. I am also not sure why this binary isn't already built for me because I've been testing using `start-services.sh` using the same polkadot release.

Edit: I see now `relaychain_bin="$relaychain_dir/target/release/polkadot"` is looking for a polkadot release in the project dir. I think we should rather respect the [env var declared here](https://github.com/Snowfork/snowbridge/pull/736/files#diff-0c324c0eec4a2767b6966f3ebbee515651fc454d364f9f3429813a162739e344L3), since if you run multiple projects you only need one polkadot instance on your machine.

File: core/packages/test/.envrc-example
Comment (claravanstaden) [inline core/packages/test/.envrc-example:3]: I don't think we should remove this, it's useful to have the polkadot instance separate from the project.

Comment (alistair-singh) [reply core/packages/test/.envrc-example:3]: +1 for this comment. I find it useful to have `polkadot` repo outside of the project repo. It is easier to clean and rebuild the project this way. Although the changes below which fetch the latest polkadot on demand could be useful for test runners.

File: core/packages/test/.envrc-example
Comment (alistair-singh) [inline core/packages/test/.envrc-example:5]: Nit: reformatting these environment variables doesn't make the code neater.

File: core/packages/test/scripts/build-binary.sh
Comment (alistair-singh) [inline core/packages/test/scripts/build-binary.sh:13]: Fetching tags and checking out branches should be inside the `if` after line 9. We should only do this on the first time initialising the parachain repo. This way we can re-run build_relaychain if we have made modifications to the polkadot repo without the `git checkout` failing.

**Comment by yrong** [comment]
> 



> @yrong Do you have an M1 Macbook by any chance? I want to see if it is possible to get these scripts running on MacOS, so we can test without having to run a separate Ubuntu VM / Instance.
> 
> If not, we can try handle the macbook issue in another PR.

@vgeddes tweak a little for mac m1 and tested in my laptop, should work

**Comment by claravanstaden** [comment]
@yrong I'm busy testing this on my Mac too, thanks!! This will be super useful.

I do think it is necessary to address [this](https://github.com/Snowfork/snowbridge/pull/736#discussion_r1044230485) first before merging, because it's useful to have Polkadot outside the repo and managed separately. For example, I cloned your Snowbridge project and I'd like to point to my existing Polkadot instance instead of checking out Polkadot again and rebuilding from scratch.

**Comment by yrong** [comment]
> @yrong I'm busy testing this on my Mac too, thanks!! This will be super useful.
> 
> I do think it is necessary to address [this](https://github.com/Snowfork/snowbridge/pull/736#discussion_r1044230485) first before merging, because it's useful to have Polkadot outside the repo and managed separately. For example, I cloned your Snowbridge project and I'd like to point to my existing Polkadot instance instead of checking out Polkadot again and rebuilding from scratch.

@claravanstaden  yes, agree and some change to respect the previous env if configured

File: core/packages/test/scripts/build-binary.sh
Comment (yrong) [reply core/packages/test/scripts/build-binary.sh:67]: AFAIK multiple runtimes in one binary is possible, not sure if we can inspect the type/feature from binary directly,  please let me know if possible. 
Except for this one I provide another script [here] (https://github.com/yrong/snowbridge/blob/refactor%2Fsetup-scripts/core/packages/test/scripts/rebuild-binary.sh) just rebuild binary from source without any check, we can manually run this script if nessessary.

**Comment by yrong** [comment]
> When I run the script, it gets stuck on deploying the Solidity contracts:
> 
> ```
> Successfully generated 104 typings!
> Compiled 48 Solidity files successfully
> deploying "Bitfield" (tx: 0x12eb8e2904b95d199358659adcd07fbb6a0452eea3c58be5e3f764cd71c79107)...
> ```
> 
> It just stays here and doesn't progress. Please let me know if you see the same on local Ethereum net?

@claravanstaden  met this issue before when I enable global proxy on my laptop and resolved after I disable it, not sure if that is the case in your env.

File: core/packages/test/scripts/build-binary.sh
Comment (claravanstaden) [reply core/packages/test/scripts/build-binary.sh:67]: I think it is possible to have multiple runtime in one binary, perhaps we should rather do that then.

I thought there was a way to inspect features from the binary directly, but I can't find it, maybe I was thinking of another feature.

**Comment by claravanstaden** [comment]
> @claravanstaden met this issue before when I enable global proxy on my laptop and resolved after disable, not sure if that is the case in your env.

@yrong I've sent you a message on Discord, can we chat about this?

**Comment by yrong** [comment]
@vgeddes @claravanstaden 

is the 60s wait here nessessary?

https://github.com/Snowfork/snowbridge/blob/bbf2b2fb67119bf2a7763d97870ce6ab5c906150/core/packages/test/scripts/start-services.sh#L380-L382
 
seems in `configure-beefy.ts` we 've also wait `beefyStartBlock` to block 29, would assume that is enough so I remove the previous wait, correct me if I'am wrong

**Comment by vgeddes** [comment]
> @vgeddes @claravanstaden 
> 
> 
> 
> is the 60s wait here nessessary?
> 
> 
> 
> https://github.com/Snowfork/snowbridge/blob/bbf2b2fb67119bf2a7763d97870ce6ab5c906150/core/packages/test/scripts/start-services.sh#L380-L382
> 
>  
> 
> seems in `configure-beefy.ts` we 've also wait `beefyStartBlock` to block 29, would assume that is enough so I remove the previous wait, correct me if I'am wrong
> 
> 

Hey yeah the wait can be removed. Good catch!

File: core/packages/test/scripts/build-binary.sh
Comment (yrong) [reply core/packages/test/scripts/build-binary.sh:67]: @claravanstaden could this comment be resolved considering [rebuild script](https://github.com/yrong/snowbridge/blob/refactor%2Fsetup-scripts/core/packages/test/scripts/rebuild-binary.sh)  or you still think inpect on the fly is required? I would assume manually rebuild should be fine.

**Comment by claravanstaden** [comment]
@yrong I have tested this PR today but I'm getting some strange results:

- On my AWS EC2 instance, I get this in the beefy-relay.log (and beefy relayer keeps restarting):

```
{"@timestamp":"2022-12-13T15:15:39.781364686Z","level":"info","logger":"stdlib","message":"2022/12/13 15:15:39 Connecting to ws://localhost:9944..."}
{"@timestamp":"2022-12-13T15:15:39.922507278Z","endpoint":"ws://localhost:9944","level":"info","message":"Connected to chain","metaVersion":14}
{"@timestamp":"2022-12-13T15:15:39.923317232Z","chainID":15,"endpoint":"ws://localhost:8546/","level":"info","message":"Connected to chain"}
{"@timestamp":"2022-12-13T15:15:39.925162583Z","beefyBlock":29,"level":"info","message":"Retrieved current BeefyClient state","validatorSetID":2}
{"@timestamp":"2022-12-13T15:15:39.926454534Z","error":"initialize ethereum writer: execution reverted","level":"fatal","message":"Unhandled error"}
```

- On my Mac, I get this (and parachain relayer keeps restarting):

```
{"@timestamp":"2022-12-13T14:38:14.252496+02:00","args":"[--api ws://localhost:11144 --block 0xac98db32d972d5b3d0c03c204014a8ab2663257add71dbaa0593d799ff8a9812]","level":"error","message":"Failed to query events.","name":"snowbridge-query-events","stdErr":"","stdOut":""}
{"@timestamp":"2022-12-13T14:38:14.252517+02:00","error":"scan for sync tasks bounded by BEEFY block 131: query events: exec: \"snowbridge-query-events\": executable file not found in $PATH","level":"fatal","message":"Unhandled error"}
```

**Comment by yrong** [comment]
> ```
>  exec: \"snowbridge-query-events\": executable file not found in $PATH"
> ```

seems this binary is not built successful in your env, would suggest run rebuild script mentioned above manually

**Comment by claravanstaden** [comment]
Hey @yrong! This is very useful work and I am very grateful for the effort you've put into this. At the same time, I think we need to make sure when we merge this that everything this works as expected. Some thoughts:

- `start-services.sh` should work by itself (as it does currently in `main`), without having to run additional scripts before that. Typically our team would run:

```
./scripts/start-services.sh
```
and in a separate terminal

```
pnpm test:integration test/bootstrap.js
```

Please check if this works.

- I think a readme with steps how to use these scripts are necessary. For example, if you make a change to the relayer, what scripts should you run to retest? Same with parachain changes.

**Comment by vgeddes** [comment]
> Do you have an M1 Macbook by any chance? I want to see if it is possible to get these scripts running on MacOS, so we can test without having to run a separate Ubuntu VM / Instance.
> 
> If not, we can try handle the macbook issue in another PR.

Leaving the MacOs support for another PR is still an option, to reduce the scope of work in this PR.

**Comment by yrong** [comment]
@claravanstaden tested and seems work fine
```
2022-12-14 21:40:37        API/INIT: RPC methods not decorated: basicOutboundChannel_getMerkleProof
    Bootstrap
      ✓ should transfer DOT from Substrate to Ethereum (basic channel) (60301ms)
      ✓ should transfer ETH from Ethereum to Substrate (basic channel) (184913ms)
```
previously I add some check if parachain/relay binary exist then do not recompile when start service, but think that may not be nessessary. As you mentioned for local dev we may change source code and run e2e test frequently, and consider the incremental build feature rust and golang provided rebuild will not take much time, so just remove the check to be same as what we do before this pr.

and change readme a little bit in this [commit](https://github.com/Snowfork/snowbridge/pull/736/commits/4d003a31cceceb960b77c914a414bdde1f2b8bff), as a non-native english speaker appreciated you may help me if anything wrong

**Comment by yrong** [comment]
> Leaving the MacOs support for another PR is still an option, to reduce the scope of work in this PR.

@vgeddes works in my mac m1 setup.  @claravanstaden does it work now in your machine? if hard to diagnosis/reproduce maybe we can focus on linux first. 

notice several projects in polkadot ecology like [composableFi](https://github.com/ComposableFi/composable/blob/main/.nix/devnet.nix) use [nix](https://nixos.org/) as package manage tool, maybe we can take a try for a reproducible/deterministic setup.

**Comment by claravanstaden** [comment]
@yrong I tested the latest changes. Upgrading Lodestar from `1.2.1` to `1.2.2` breaks the beacon relayer, so you'll need to revert that. It's good practice to run `start-services.sh` after every change to check if everything still works as expected, just FYI for future changes. I reverted that locally and then everything seems OK.

**Comment by yrong** [comment]
> @yrong I tested the latest changes. Upgrading Lodestar from `1.2.1` to `1.2.2` breaks the beacon relayer, so you'll need to revert that. It's good practice to run `start-services.sh` after every change to check if everything still works as expected, just FYI for future changes. I reverted that locally and then everything seems OK.

@claravanstaden are you sure bump to 1.2.2 is the root cause? tests still passed in my setup, could you attach some detail error logs?

File: core/packages/test/scripts/start-services.sh
Comment (claravanstaden) [inline core/packages/test/scripts/start-services.sh:15]:
```suggestion
echo "Installing binaries if required"
```

**Review by claravanstaden — APPROVED** [review]
@yrong I reverted Lodestar back to `1.2.2` and I'm not getting the same error again. 🤷‍♀️ I don't have the logs anymore, but it gave `404` errors on the Lodestar finalized update endpoint. Maybe it was just something temperamental with the run. So I'm happy to approve!

File: core/packages/test/package.json
Comment (alistair-singh) [inline core/packages/test/package.json:19]: @yrong Does this bump fix the issues where the testnet was not starting up? @claravanstaden would it be better to pin lodestar to a specific version? I feel like pinning to a version will make troubleshooting environment start up issues easier.

File: core/packages/test/scripts/deploy-polkadot.sh
Comment (alistair-singh) [inline core/packages/test/scripts/deploy-polkadot.sh:76]: Nit: remove testing comment and add a new line.

File: core/packages/test/scripts/start-chains.sh
Comment (alistair-singh) [inline core/packages/test/scripts/start-chains.sh:27]: Nit: same as above. Remove testing comments.

File: core/packages/test/scripts/start-relayer.sh
Comment (alistair-singh) [inline core/packages/test/scripts/start-relayer.sh:138]: Nit: testing comments

File: core/packages/test/package.json
Comment (claravanstaden) [reply core/packages/test/package.json:19]: @alistair-singh @yrong yeah good idea, let's pin to `v1.2.2` since a lot can change within a version.

Comment (yrong) [reply core/packages/test/package.json:19]: @alistair-singh 
TBO not sure it is the root cause, but in one of my test env the beacon relayer return 404 error when [query finalized update](https://github.com/Snowfork/snowbridge/blob/c16aa0d63f7450440e1529d08761acd5a65ee9dc/relayer/relays/beacon/header/syncer/api.go#L505) with v1.2.1 while upgrading to v1.2.2 resolved, so I decide try to upgrade to v1.2.2.  Have you met this issue before and would appreciated if you guys can help to verify 1.2.2 is fine, then I am ok pin to this version.

Btw, the correct response should be something like this:

```
 curl http://localhost:9596/eth/v1/beacon/light_client/finality_update
{"data":{"attested_header":{"slot":"120","proposer_index":"0","parent_root":"0x66663ce62fb88b3e87b9ad85f4b57f5c0c1dc2c17f4eee26c7cf03003d0a1929","state_root":"0xde5434d3834b130d75cf77ea36792c4cdf544b9210606993db31f9bf6329b2a9","body_root":"0x96dd609f95b8c2950478fd4d4f5a8f10ce951797f8b32fabb39e83406dd6d2c5"},"finalized_header":{"slot":"104","proposer_index":"3","parent_root":"0x8457d1b603ba7aa9f9a330445b264521f4f26a88b05610ca17b9962271d54fd4","state_root":"0x73af2312436de3c3c749a861806c8401c30acc18bb5d35458c03cdced933b4ea","body_root":"0x70f2511e36524201177eddabb5a084d53e12671812899fe8fcd9d0d49f80e95e"},"finality_branch":["0x0d00000000000000000000000000000000000000000000000000000000000000","0x10c726fac935bf9657cc7476d3cfa7bedec5983dcfb59e8a7df6d0a619e108d7","0x2d032c327de032b41addfcce3addbee0e52706df98a8c880493b8e22de19b3c1","0xcc3a4a103f88b4631f9a1819340078938e503982bb6b98d9487963b5a1d2ce20","0x67fb2b9520445ff5f9fc9d84e94796b892144365cf693e3c7eb74ef969f1f60e","0x7d80e0deb52eaa4f5f409c377cdd3a29d0f50e081edc01ad89c4926c0d72aa97"],"sync_aggregate":{"sync_committee_bits":"0xffffffff","sync_committee_signature":"0xaa55cb9b92e23d995fb41e33c637b1ff45aa35802f2853253a8c6a8b3944798b93ae92d68257c03d99dfae400960893f05ba2207d37e6ba4240ebfb327927d267238ddfb9f81004ca886d03b062e379aaf2205852510f9c6349c61f9daf06c79"},"signature_slot":"121"}}
```

**Review by doubledup — COMMENTED** [review]
Hey, thanks for this! Haven't looked through everything yet, but this is my feedback so far

File: core/packages/test/package.json
Comment (claravanstaden) [reply core/packages/test/package.json:19]: Interesting @yrong! It seems like the same issue. I'm happy to pin on 1.2.2 and let's keep an eye on that issue if we see it again.

**Review by alistair-singh — COMMENTED** [review]
I have tested this on Ubuntu and everything works. Just waiting on some minor fixes before I approve.

**Comment by yrong** [comment]
> I have tested this on Ubuntu and everything works. Just waiting on some minor fixes before I approve.

@alistair-singh please help to check this [commit](https://github.com/Snowfork/snowbridge/pull/736/commits/ac903c6a1f13eb7c20cc4191b2c1de1929949d90) which resolves comments above. Except that one change introduced is to allow starting chains and relayer seperately, in this mode we can focus debuging relayer without restarting/initializing chains which take much time, hope that is helpful.

File: core/packages/test/scripts/start-relayer.sh
Comment (yrong) [reply core/packages/test/scripts/start-relayer.sh]: https://github.com/Snowfork/snowbridge/pull/736#issuecomment-1354052706

File: core/packages/test/scripts/start-chains.sh
Comment (yrong) [reply core/packages/test/scripts/start-chains.sh]: https://github.com/Snowfork/snowbridge/pull/736#issuecomment-1354052706

**Review by alistair-singh — APPROVED** [review]
Tested! 🚀

**Pull request merged**
